src/plugins/biostruct3d_view/src/BioStruct3DGLWidget.cpp

```diff
@@ -967,6 +965,7 @@ void BioStruct3DGLWidget::sl_showSurface() {
 
     QString surfaceType = qobject_cast<QAction*>(sender())->text();
     surfaceCalcTask = new MolecularSurfaceCalcTask(surfaceType, atoms);
+    connect(surfaceCalcTask, &Task::si_stateChanged, this, &BioStruct3DGLWidget::sl_onTaskFinished);
```

Comment:
This is very unsafe to keep 'surfaceCalcTask' globally because it is automatically deleted by the Task Scheduler when it is finished. This way we have a broken pointer cached in the 'surfaceCalcTask'.

I would recommend changing 'surfaceCalcTask' to a local variable and using sl_onTaskFinished(Task* task) 'task' parameter inside of the sl_onTaskFinished to restore it value.

The sl_onTaskFinished should also get a better name like sl_onSurfaceCalcTaskFinished

Reply:
> This is very unsafe

It's wrapped in a QPointer.

> using sl_onTaskFinished(Task* task) 'task' parameter inside of the sl_onTaskFinished to restore it value.

In the previous implementation, the `sl_onTaskFinished` signal triggered for all `TaskScheduler::si_stateChanged` signals from all tasks (not just `surfaceCalcTask`) during the lifetime of the BioStruct3DGLWidget (not just `MolecularSurfaceCalcTask`). The `sl_onTaskFinished(Task*)` slot signature will not allow it to be connected to the **Task**::si_stateChanged() signal (only to **TaskScheduler**::si_stateChanged(Task *)), which means that it will not be possible to limit the "scope" of the slot and the "lifetime" of the slot.

Getting rid of the pointer is a good idea, but then I think it's better to use `sender()` and cast to `MolecularSurfaceCalcTask` in the current implementation instead of using a slot with signature `sl_onTaskFinished(Task* task)`.